Create a skeleton controller

ncskier opened the pull request:
# Changes

<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

Create the skeleton controller described in issue https://github.com/tektoncd/triggers/issues/8.

- Add skeleton reconciler files for the `eventlistener`, `triggertemplate`, and `triggerbinding`
- Add config files for the controller to get a deployment going with `ko apply -f config/`.
- Add an e2e test with `ko apply -f config/`
- Add the `hack/update-deps.sh` script and a ton of `vendor/` dependencies

~#### WIP:~
- ~I ran into some issues with `dep`, discussed on Slack [here](https://tektoncd.slack.com/archives/GKV9LUVHV/p1562360044024800). The code in this PR builds correctly right now, but I would like to get some feedback about this issue.~
I've added extra constraints to the `Gopkg.toml` file to solve this issue.
- ~There is placeholder code commented out right now that depends on the generated clientset & informer code that is being added under PR https://github.com/tektoncd/triggers/pull/17.~
The PR has been merged, and I created a reconciler for the `EventListener`.
- ~The `triggers` controller Deployment uses the [`tekton-pipelines-controller`](https://github.com/tektoncd/pipeline/blob/master/config/200-serviceaccount.yaml) ServiceAccount. Under issue https://github.com/tektoncd/triggers/issues/7, this can be changed to a triggers ServiceAccount if we do not want to depend on `pipelines` being installed.~
With PR https://github.com/tektoncd/triggers/pull/23 we use our own `triggers` Service Account.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

vdemeester commented:
@ncskier might be worth exploring *injection-based* controllers (see https://github.com/tektoncd/pipeline/pull/1015 as a reference)

bobcatfish commented:
might as well switch to the injection based controller right at the outset if we can! (sorry @ncskier 😅 )

bobcatfish commented on config/controller-service.yaml:
```diff
@@ -0,0 +1,29 @@
+  - name: metrics
+    port: 9090
+    protocol: TCP
+    targetPort: 9090
```
we have been copying this 9090 port thing around (from knative) - do we even need it? I think it might be for emitting metrics that I'm not sure we're actually emitting? (or maybe the knative libs are?)

(just ignore me if you'd rather not look into this)

ncskier replied:
Looks like it's for knative/pkg/metrics, although I'm not sure if we're emitting these or not:
https://github.com/tektoncd/pipeline/blob/master/vendor/github.com/knative/pkg/metrics/config.go#L56

bobcatfish replied:
nice! thanks for tracking that down :D I've been wondering for so long XD

bobcatfish commented on pkg/reconciler/reconciler.go:
```diff
@@ -0,0 +1,137 @@
+type Options struct {
+	KubeClientSet     kubernetes.Interface
+	SharedClientSet   sharedclientset.Interface
+	PipelineClientSet pipelineclientset.Interface
```
oh nice, so we'll have this as a vendored dep already 👍

bobcatfish commented on test/e2e-tests.sh:
```diff
@@ -0,0 +1,42 @@
+header "Running Go e2e tests"
+go_test_e2e -timeout=20m ./test || fail_test
+
+success
```
woot thanks for adding the end to end test boilerplate!

ncskier commented:
I'll take a look 😅

ncskier commented:
@bobcatfish @vdemeester I've updated the controller to use the knative/pkg injection infrastructure. Please take another look at reviewing this PR when you get the chance 🙂 

I only added a reconciler for the `EventListener` at the moment, because I'm not sure if we'll need reconcilers for the other types (and if we need a reconciler we can easily add one); let me know if you have an opinion about this.

I also changed the bash scripts to use `tektoncd/plumbing` instead of `test-infra`.

bobcatfish commented on .golangci.yml:
```diff
@@ -0,0 +1,9 @@
+  - pkg/client/clientset/versioned/fake
+linters-settings:
+  errcheck:
+    exclude: .errcheck.txt
```
ooo what's this file for? ive never seen this before! :D

(question about both the .golangci.yaml and the .errcheck.txt!)

ncskier replied:
This is the config file for our linter `golangci-lint` (more info [here](https://github.com/golangci/golangci-lint#config-file)). `.errcheck.txt` lists files to exclude from testing. However, I don't think we need `.errcheck.txt` anymore, because the linter error we were skipping was fixed [the correct way](https://github.com/tektoncd/triggers/pull/21#discussion_r305095390) in @vincent-pli's webhook PR 😅 

I'm going to remove the `.errcheck.txt` file in a PR that I'm putting together, because I don't think we need it anymore.

bobcatfish commented on config/config-logging.yaml:
```diff
@@ -0,0 +1,50 @@
+  name: config-logging-triggers
+  namespace: tekton-pipelines
+data:
+  # Common configuration for all knative codebase
```
comment out of date haha

bobcatfish commented on config/config-logging.yaml:
```diff
@@ -0,0 +1,50 @@
+
+  # Log level overrides
+  loglevel.controller: "info"
+  loglevel.webhook: "info"
```
seems like the level is already at "info", i wonder why these are explicitly also setting info 🤔

bobcatfish commented on config/config-observability.yaml:
```diff
@@ -0,0 +1,53 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: config-observability
```
interesting, is this configuring functionality we get out of the box with knative/pkg?

ncskier replied:
I think so... @vdemeester do you know about the `config-observability` functionality? 😄

bobcatfish commented on hack/update-codegen.sh:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-# Copyright 2018 The Knative Authors
+# Copyright 2019 The Tekton Authors
```
hee hee
